kernel/kernel-api/src/main/java/io/delta/kernel/internal/checkpoints/CheckpointInstance.java

```diff
@@ -117,14 +168,26 @@ public boolean equals(Object o) {
         if (o == null || getClass() != o.getClass()) {
             return false;
         }
+        // For V2 checkpoints, compare the filepath.
         CheckpointInstance checkpointInstance = (CheckpointInstance) o;
+        if (checkpointInstance.format == CheckpointFormat.V2 && format == CheckpointFormat.V2 &&
+                !filePath.equals(checkpointInstance.filePath)) {
+            return false;
+        }
         return version == checkpointInstance.version &&
-                Objects.equals(numParts, checkpointInstance.numParts);
+                Objects.equals(numParts, checkpointInstance.numParts) &&
+                format == checkpointInstance.format;
     }
 
     @Override
     public int hashCode() {
-        return Objects.hash(version, numParts);
+        // For V2 checkpoints, include the filepath in the hash of the instance (as we consider
+        // different UUID checkpoints to be different checkpoint instances. Otherwise, ignore
+        // the filepath when hashing.
+        if (format == CheckpointFormat.V2) {
+            return Objects.hash(version, numParts, format, filePath);
+        }
+        return Objects.hash(version, numParts, format, "");
```

Comment:
why not use include the filePath always and without the if and else.

Reply:
Fixed.

Reply:
Actually, we need to exclude filePath (to consider multipart checkpoints equal if they have different paths).

Reply:
Please see my other comment.